File: src/features/projects/dto/create-project.dto.ts
import { IsArray, IsObject, IsOptional, IsString } from 'class-validator';

export class CreateProjectDto {
  @IsString()
  name: string;

  @IsOptional()
  @IsString()
  description?: string;

  @IsOptional()
  @IsString()
  orgId?: string;

  @IsOptional()
  @IsString()
  officeId?: string;

  @IsOptional()
  @IsString()
  projectCode?: string;

  @IsOptional()
  @IsString()
  status?: string;

  @IsOptional()
  @IsString()
  location?: string;

  @IsOptional()
  @IsString()
  bidDate?: string;

  @IsOptional()
  @IsString()
  awardDate?: string;

  @IsOptional()
  @IsString()
  fabricationStartDate?: string;

  @IsOptional()
  @IsString()
  fabricationEndDate?: string;

  @IsOptional()
  @IsString()
  erectionStartDate?: string;

  @IsOptional()
  @IsString()
  erectionEndDate?: string;

  @IsOptional()
  @IsString()
  completionDate?: string;

  @IsOptional()
  @IsObject()
  budget?: Record<string, any>;

  @IsOptional()
  @IsObject()
  quantities?: Record<string, any>;

  @IsOptional()
  @IsObject()
  staffing?: Record<string, any>;

  @IsOptional()
  @IsArray()
  costCodeBudgets?: Record<string, any>[];
}

File: spring-boot-api/src/main/java/com/mytegroup/api/dto/projects/CreateProjectDto.java
package com.mytegroup.api.dto.projects;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;

public record CreateProjectDto(
    String name,
    String description,
    String orgId,
    String officeId,
    String projectCode,
    String status,
    String location,
    LocalDate bidDate,
    LocalDate awardDate,
    LocalDate fabricationStartDate,
    LocalDate fabricationEndDate,
    LocalDate erectionStartDate,
    LocalDate erectionEndDate,
    LocalDate completionDate,
    Map<String, Object> budget,
    Map<String, Object> quantities,
    Map<String, Object> staffing,
    List<Map<String, Object>> costCodeBudgets
) {
    public CreateProjectDto {
        if (name != null) {
            name = name.trim();
        }
        if (description != null) {
            description = description.trim();
        }
        if (projectCode != null) {
            projectCode = projectCode.trim();
        }
        if (status != null) {
            status = status.trim();
        }
        if (location != null) {
            location = location.trim();
        }
    }
}



